### PR TITLE
Deprecate ElementGroupBase.index

### PR DESCRIPTION
**Pull request by alexfikl**
This deprecates the `index` attribute
* Set `index=None` in `ElementGroupBase` and issue a deprecation warning.
* Set `index=None` in `ElementGroupFactory.__call__` too.

The last two commits are independent and just do some random cleanup in `meshmode.discretization.__init__`. I can pluck those out into a separate PR if that's easier.

Second part of #224.

### CODE REVIEW

File: meshmode/discretization/__init__.py [hunk 1]
```diff
@@ -101,11 +107,15 @@ class ElementGroupBase(metaclass=ABCMeta):
...
-            :class:`~meshmode.mesh.MeshElementGroup`.
-        """
+    def __init__(self,
+                 mesh_el_group: _MeshElementGroup,
+                 order: int,
+                 index: Optional[int] = None) -> None:
+        if index is not None:
+            warn("Setting 'index' is deprecated and will be removed in 2023. "
+                 "This is now handled internally by the discretization.",
+                 DeprecationWarning, stacklevel=2)
```
Comment (alexfikl) [inline meshmode/discretization/__init__.py:117]: Would it be better to raise the warning only when someone tries to use `grp.index`? Here it's easy to wrap it up in a property and do that.

File: meshmode/discretization/__init__.py [hunk 2]
```diff
@@ -125,7 +135,8 @@ def nelements(self):
         """
         return self.mesh_el_group.nelements
 
-    @abstractproperty
+    @property
+    @abstractmethod
```
Comment (inducer) [inline meshmode/discretization/__init__.py:139]: Why this change? (also below)

File: meshmode/discretization/__init__.py [hunk 3]
```diff
@@ -172,6 +185,20 @@ def discretization_key(self):
         """
         return (type(self), self.dim, self.order)
 
+
+class ElementGroupFactory:
```
Comment (inducer) [inline meshmode/discretization/__init__.py:189]: Make this a protocol (cf. `typing.Protocol`)? (in preparation for typed `meshmode`) The intention from my end was very much that a plain callable can be a element group factory.

File: meshmode/discretization/__init__.py [hunk 4]
```diff
@@ -396,26 +424,31 @@ class Discretization:
...
+                 mesh: _Mesh,
+                 group_factory: ElementGroupFactory,
+                 real_dtype: Optional[np.dtype] = None,
+                 _force_actx_clone: bool = True) -> None:
         """
-        :arg actx: A :class:`ArrayContext` used to perform computation needed
-            during initial set-up of the mesh.
-        :arg mesh: A :class:`meshmode.mesh.Mesh` over which the discretization is
+        :arg actx: an :class:`arraycontext.ArrayContext` used to perform
+            computation needed during initial set-up of the mesh.
```
Comment (inducer) [inline meshmode/discretization/__init__.py:434]:
```suggestion
            computation needed during initial set-up of the discretization.
```

**Review by inducer — COMMENTED** [review]
Thanks for working on this. Just a few things below.

File: meshmode/discretization/__init__.py (hunk 3 above)
Comment (alexfikl) [reply meshmode/discretization/__init__.py]: This found a bug too!
https://github.com/inducer/meshmode/compare/3f463c49f29f1ea436ca9b05cf42e285bcce62b3..9797d4775cde156d2bc3de883adfdd9aba9f778b

File: meshmode/discretization/__init__.py (hunk 2 above)
Comment (inducer) [reply meshmode/discretization/__init__.py:147]: Huh, TIL. Thanks!

**Comment by inducer** [comment]
Thanks!
